linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java

```diff
@@ -1803,7 +1803,21 @@ private boolean innerHasNext() {
       final Predicate2<TSource, TInner> predicate,
       Function2<TSource, TInner, TResult> resultSelector,
       final JoinType joinType) {
-    // Building the result as a list is easy but hogs memory. We should iterate.
+    if (!joinType.generatesNullsOnLeft()) {
+      return nestedLoopJoinOptimized(outer, inner, predicate, resultSelector, joinType);
+    }
+    return nestedLoopJoinAsList(outer, inner, predicate, resultSelector, joinType);
+  }
+
+  /**
+   * Implementation of nested loop join that builds the complete result as a list
+   * and then returns it. This is an easy-to-implement solution, but hogs memory.
+   */
+  private static <TSource, TInner, TResult> Enumerable<TResult> nestedLoopJoinAsList(
+      final Enumerable<TSource> outer, final Enumerable<TInner> inner,
+      final Predicate2<TSource, TInner> predicate,
+      Function2<TSource, TInner, TResult> resultSelector,
+      final JoinType joinType) {
```

Comment:
`generateNullsOnLeft` will always be true, isn't it ?

Reply:
And the join type will never be `ANTI` or `SEMI`

Reply:
At this point, yes. 
However, I prefer to keep this method as it is, supporting all type of joins, so that we can always have it as a "fallback", in case we find any issues with the new algorithm.

Reply:
Personally, I would prefer to use git history as the "fallback", the code here is a little redundant, and may be confusing.

Reply:
I understand your point, and I'm not against it; but I'd rather keep the join algorithms as "complete" as possible. I don't think leaving this method here as it is will be confusing. Let's see what the rest of the community think about this topic.

Reply:
By the way, {{class EnumerableDefaults}} exists because a java before JDK 8 did not allow default implementations of interface methods.

Now that has changed we could consider removing {{EnumerableDefaults}}.

Reply:
+1 to keep all the join types.

Reply:
@julianhyde , you're right, I have created a separate ticket [1] to track that refactoring.

[1] https://issues.apache.org/jira/browse/CALCITE-3555